flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/enumerator/tracker/SplitTracker.java

```diff
@@ -0,0 +1,237 @@
+     * not there, that means that that should already be cleaned up.
+     */
+    public List<DynamoDbStreamsShardSplit> splitsAvailableForAssignment() {
+        return knownSplits.values().stream()
+                .filter(
+                        split -> {
+                            boolean splitIsNotAssigned = !isAssigned(split.splitId());
+                            return splitIsNotAssigned
+                                    && !isFinished(split.splitId())
+                                    && (verifyAllParentSplitsAreFinished(split)
+                                            || verifyAllParentSplitsAreCleanedUp(split));
+                        })
+                .collect(Collectors.toList());
+    }
+
+    public List<DynamoDBStreamsShardSplitWithAssignmentStatus> snapshotState(long checkpointId) {
+        return knownSplits.values().stream()
+                .map(
+                        split -> {
+                            SplitAssignmentStatus assignmentStatus =
+                                    SplitAssignmentStatus.UNASSIGNED;
+                            if (isAssigned(split.splitId())) {
+                                assignmentStatus = ASSIGNED;
+                            } else if (isFinished(split.splitId())) {
+                                assignmentStatus = SplitAssignmentStatus.FINISHED;
+                            }
+                            return new DynamoDBStreamsShardSplitWithAssignmentStatus(
+                                    split, assignmentStatus);
+                        })
+                .collect(Collectors.toList());
+    }
+
+    @VisibleForTesting
+    public Set<String> getKnownSplitIds() {
+        return knownSplits.keySet();
+    }
+
+    /**
+     * finishedSplits needs to be cleaned up. The logic applied to cleaning up finished splits is
+     * that if any split has been finished reading, its parent has been finished reading and it is
+     * no longer present in the describestream response, it means that the split can be cleaned up.
+     */
+    public void removeSplits(Set<String> discoveredSplitIds) {
+        Set<String> finishedSplitsSnapshot = new HashSet<>(finishedSplits);
+        for (String finishedSplitId : finishedSplitsSnapshot) {
+            DynamoDbStreamsShardSplit finishedSplit = knownSplits.get(finishedSplitId);
+            if (isSplitReadyToBeCleanedUp(finishedSplit, discoveredSplitIds)) {
+                finishedSplits.remove(finishedSplitId);
+                knownSplits.remove(finishedSplitId);
+            }
+        }
+    }
+
+    private boolean isSplitReadyToBeCleanedUp(
+            DynamoDbStreamsShardSplit finishedSplit, Set<String> discoveredSplitIds) {
+        String splitId = finishedSplit.splitId();
+        boolean parentSplitsAreFinished = verifyAllParentSplitsAreFinished(finishedSplit);
+        boolean parentSplitsAreCleanedUp = verifyAllParentSplitsAreCleanedUp(finishedSplit);
```

Comment:
why do we care about parents? Might cause issues when we restore from stale snapshot (all shards are expired and should be cleaned up)

Reply:
we only want to clean a split up if its parents are read or have been cleaned up. For a restore from stale snapshot, we'll be assigning the splits and GetRecords or GetShardIterator will mark the splits as finished and they'll get put in finished splits.